semantic_search/main.py

```diff
@@ -71,24 +83,39 @@ def app_startup():
     model.tokenizer, model.model = setup_model_and_tokenizer(
         settings.pretrained_model_name_or_path, cuda_device=settings.cuda_device
     )
-    model.similarity = torch.nn.CosineSimilarity(-1)
+    embedding_dim = model.model.config.hidden_size
+    model.index = setup_faiss_index(embedding_dim)
 
 
 @app.post("/")
 async def query(query: Query) -> List[Dict[str, float]]:
-    ids = [query.query.uid] + [document.uid for document in query.documents]
-    text = [query.query.text] + [document.text for document in query.documents]
-
-    embeddings = encode(text)
-    similarity_scores = model.similarity(embeddings[0], embeddings[1:])
-
-    # If top_k not specified, return all documents.
-    top_k = similarity_scores.size(0)
-    if query.top_k is not None:
-        top_k = max(min(query.top_k, len(query.documents)), 0)
-    top_k_scores, top_k_indicies = torch.topk(similarity_scores, top_k)
-    top_k_scores = top_k_scores.tolist()
-    # Offset the indices by 1 to account for the query
-    top_k_indicies = [idx.item() + 1 for idx in top_k_indicies]
-
-    return [{"uid": ids[idx], "score": top_k_scores[num]} for num, idx in enumerate(top_k_indicies)]
+    ids = [int(doc.uid) for doc in query.documents]
+    texts = [document.text for document in query.documents]
+
+    # # Ensure that the query is not in the index when we search.
+    # query_id = np.asarray(int(query.query.uid)).reshape(
+    #     1,
+    # )
+    # model.index.remove_ids(query_id)
+
+    # Only add items to the index if they do not already exist.
+    # See: https://github.com/facebookresearch/faiss/issues/859
+    # To do this, we first determine which of the incoming ids do not exist in the index
+    indexed_ids = set(faiss.vector_to_array(model.index.id_map).tolist())
+    to_embed = [(id_, text) for id_, text in zip(ids, texts) if id_ not in indexed_ids]
+    # We then embed the corresponding text and update the index
+    if to_embed:
+        ids, texts = zip(*to_embed)  # type: ignore
+        embeddings = encode(texts).cpu().numpy()
+        add_to_faiss_index(ids, embeddings, model.index)
+
+    top_k = max(min(query.top_k, len(query.documents)), 0) if query.top_k else None
```

Comment:
I am not sure this makes a ton of sense anymore. With the index, I think we can drop it. If `query.top_k == None` or `query.top_k > len(index)`, FAISS is smart enough to just return everything in the index within the search results, which seems like a sensible default to me. @jvwong is this okay?

We could move validation (e.g. checking that `query.top_k > 0` further upstream.

Reply:
After discussion with @maxkfranz, decided to go with `min(index.ntotal, 10)`.